File: android/designsystem/src/main/kotlin/com/bolsillo/designsystem/component/MoneyText.kt
package com.bolsillo.designsystem.component

import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import com.bolsillo.designsystem.theme.BolsilloTheme
import java.text.NumberFormat
import java.util.Locale
import kotlin.math.abs

/**
 * Single money renderer used everywhere a monetary amount is displayed.
 *
 * Takes PRIMITIVES (not the domain `Money` value class) so :designsystem stays
 * dependency-free of :domain (Article VIII / parity with iOS).
 *
 * - `minorUnits` is the signed amount; negative renders as `−$X`, positive as `$X`.
 * - `decimalDigits` comes from `Currency.decimalDigits` (USD=2, COP=2, ...).
 * - Locale drives the grouping/decimal separator. The number formatter NEVER
 *   touches `Double` — integer math splits into integer/fraction parts.
 * - `tabular figures` are guaranteed by the typography style (fontFeatureSettings).
 */
@Composable
fun MoneyText(
    minorUnits: Long,
    decimalDigits: Int,
    symbol: String,
    modifier: Modifier = Modifier,
    locale: Locale = Locale.getDefault(),
    style: TextStyle = BolsilloTheme.typography.amountRow,
    forceSign: Boolean = false,
    color: Color = signColor(minorUnits, forceSign),
) {
    Text(
        text = formatMoney(minorUnits, decimalDigits, symbol, locale, forceSign),
        modifier = modifier,
        style = style,
        color = color,
    )
}

@Composable
private fun signColor(
    minorUnits: Long,
    forceSign: Boolean,
): Color {
    val colors = BolsilloTheme.colors
    return when {
        minorUnits < 0L -> colors.amountNegative
        minorUnits > 0L && forceSign -> colors.amountPositive
        else -> colors.textPrimary
    }
}

internal fun formatMoney(
    minorUnits: Long,
    decimalDigits: Int,
    symbol: String,
    locale: Locale,
    forceSign: Boolean,
): String {
    val magnitude = abs(minorUnits)
    val divisor = pow10(decimalDigits)
    val whole = magnitude / divisor
    val fraction = magnitude % divisor
    val nf = NumberFormat.getIntegerInstance(locale)
    val wholeStr = nf.format(whole)
    val decimalSep = decimalSeparator(locale)
    val body =
        if (decimalDigits == 0) {
            wholeStr
        } else {
            val fractionStr = fraction.toString().padStart(decimalDigits, '0')
            "$wholeStr$decimalSep$fractionStr"
        }
    val prefix =
        when {
            minorUnits < 0L -> "−$symbol"
            minorUnits > 0L && forceSign -> "+$symbol"
            else -> symbol
        }
    return "$prefix$body"
}

private fun pow10(digits: Int): Long {
    var v = 1L
    repeat(digits) { v *= 10L }
    return v
}

private fun decimalSeparator(locale: Locale): String {
    val sym =
        (NumberFormat.getInstance(locale) as? java.text.DecimalFormat)
            ?.decimalFormatSymbols?.decimalSeparator ?: '.'
    return sym.toString()
}

File: ios/Packages/Bolsillo/Sources/BolsilloDesignSystem/Components/MoneyText.swift
import SwiftUI

/// Renders a money amount from primitive inputs only — no domain Money import (Article VIII).
/// `minorUnits` is the integer amount in the currency's smallest unit (e.g. cents).
/// `decimalDigits` is the number of decimal places for the currency (e.g. 2 for USD).
/// Display conversion uses integer arithmetic only — never floating-point.
public struct MoneyText: View {
    public let minorUnits: Int
    public let decimalDigits: Int
    public let symbol: String
    public let locale: Locale
    public let showSign: Bool

    @Environment(\.bolsilloTheme) private var theme

    public init(
        minorUnits: Int,
        decimalDigits: Int,
        symbol: String,
        locale: Locale = .current,
        showSign: Bool = false
    ) {
        self.minorUnits    = minorUnits
        self.decimalDigits = decimalDigits
        self.symbol        = symbol
        self.locale        = locale
        self.showSign      = showSign
    }

    public var body: some View {
        Text(displayString)
            .font(theme.typography.amountRow)
            .foregroundStyle(amountColor)
    }

    // MARK: Private helpers

    private var amountColor: Color {
        minorUnits >= 0 ? theme.colors.amountPositive : theme.colors.amountNegative
    }

    private var displayString: String {
        let absUnits = abs(minorUnits)
        let formatted: String
        if decimalDigits == 0 {
            formatted = "\(symbol)\(absUnits)"
        } else {
            let divisor   = pow10(decimalDigits)
            let whole     = absUnits / divisor
            let remainder = absUnits % divisor
            let fracStr   = String(remainder).leftPadded(toLength: decimalDigits, with: "0")
            let decSep    = locale.decimalSeparator ?? "."
            formatted = "\(symbol)\(whole)\(decSep)\(fracStr)"
        }
        if showSign {
            return minorUnits < 0 ? "-\(formatted)" : "+\(formatted)"
        }
        return minorUnits < 0 ? "-\(formatted)" : formatted
    }

    private func pow10(_ n: Int) -> Int {
        var result = 1
        for _ in 0..<n { result *= 10 }
        return result
    }
}

private extension String {
    func leftPadded(toLength length: Int, with pad: Character) -> String {
        guard count < length else { return self }
        return String(repeating: pad, count: length - count) + self
    }
}
